integration-tests/src/test/java/com/datastax/oss/driver/mapper/QueryProviderIT.java

```diff
@@ -0,0 +1,224 @@
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.PagingIterable;
+import com.datastax.oss.driver.api.core.cql.BoundStatementBuilder;
+import com.datastax.oss.driver.api.core.cql.PreparedStatement;
+import com.datastax.oss.driver.api.core.cql.SimpleStatement;
+import com.datastax.oss.driver.api.mapper.MapperContext;
+import com.datastax.oss.driver.api.mapper.annotations.ClusteringColumn;
+import com.datastax.oss.driver.api.mapper.annotations.Dao;
+import com.datastax.oss.driver.api.mapper.annotations.DaoFactory;
+import com.datastax.oss.driver.api.mapper.annotations.DaoKeyspace;
+import com.datastax.oss.driver.api.mapper.annotations.Entity;
+import com.datastax.oss.driver.api.mapper.annotations.Insert;
+import com.datastax.oss.driver.api.mapper.annotations.Mapper;
+import com.datastax.oss.driver.api.mapper.annotations.PartitionKey;
+import com.datastax.oss.driver.api.mapper.annotations.QueryProvider;
+import com.datastax.oss.driver.api.mapper.entity.EntityHelper;
+import com.datastax.oss.driver.api.querybuilder.select.Select;
+import com.datastax.oss.driver.api.testinfra.ccm.CcmRule;
+import com.datastax.oss.driver.api.testinfra.session.SessionRule;
+import com.datastax.oss.driver.categories.ParallelizableTests;
+import java.util.Objects;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TestRule;
+
+@Category(ParallelizableTests.class)
+public class QueryProviderIT {
+
+  private static CcmRule ccm = CcmRule.getInstance();
+  private static SessionRule<CqlSession> sessionRule = SessionRule.builder(ccm).build();
+  @ClassRule public static TestRule chain = RuleChain.outerRule(ccm).around(sessionRule);
+
+  private static SensorDao dao;
+
+  @BeforeClass
+  public static void setup() {
+    CqlSession session = sessionRule.session();
+
+    session.execute(
+        SimpleStatement.builder(
+                "CREATE TABLE sensor_reading(sensor_id int, month int, day int, value double, "
+                    + "PRIMARY KEY (sensor_id, month, day)) "
+                    + "WITH CLUSTERING ORDER BY (month DESC, day DESC)")
+            .setExecutionProfile(sessionRule.slowProfile())
+            .build());
+
+    SensorMapper mapper = new QueryProviderIT_SensorMapperBuilder(session).build();
+    dao = mapper.sensorDao(sessionRule.keyspace());
+  }
+
+  @Test
+  public void should_invoke_query_provider() {
+    SensorReading readingFeb3 = new SensorReading(1, 2, 3, 9.3);
+    SensorReading readingFeb2 = new SensorReading(1, 2, 2, 8.6);
+    SensorReading readingFeb1 = new SensorReading(1, 2, 1, 8.7);
+    SensorReading readingJan31 = new SensorReading(1, 1, 31, 8.2);
+    dao.save(readingFeb3);
+    dao.save(readingFeb2);
+    dao.save(readingFeb1);
+    dao.save(readingJan31);
+
+    assertThat(dao.findSlice(1, null, null).all())
```

Comment:
Wish java had optional arguments :D.  I guess technically you could overload this in the Dao interface and in `FindSliceProvider`

Reply:
Yes.

Actually for this particular case it would probably be more simple to declare 3 `@Query` methods, but that's just a silly example.